packages/react-dom/src/client/ReactDOMComponentTree.js

```diff
@@ -47,11 +47,35 @@ export function detachDeletedInstance(node: Instance): void {
   // these fields are relevant.
   delete (node: any)[internalInstanceKey];
   delete (node: any)[internalPropsKey];
+  delete (node: any)[internalContainerInstanceKey];
   delete (node: any)[internalEventHandlersKey];
   delete (node: any)[internalEventHandlerListenersKey];
   delete (node: any)[internalEventHandlesSetKey];
 }
 
+function _tagIsDetachable(tag: number): boolean {
+  if (tag == null) {
+    return false;
+  }
+  return tag === HostComponent || tag === HostText;
+}
+
+export function detatchFiberStateNode(
+  fiberToDetach: Fiber,
+) {
+  if (_tagIsDetachable(fiberToDetach.tag)) {
+    const hostInstance: Instance = fiberToDetach.stateNode;
+    if (hostInstance !== null) {
+      detachDeletedInstance(hostInstance);
+    }
+  }
+  fiberToDetach.stateNode = null;
+
+  if (fiberToDetach.alternate !== null) {
```

Comment:
From my understanding, `fiber.alternate` is the Fiber that stores future/new work. Not all fibers will have future work, but if we're unmounting a fiber, we should be sure to cleanup the "future" state too.